x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -182,6 +182,24 @@ export class AuthenticationService {
     });
 
     http.registerOnPreResponse(async (request, preResponse, toolkit) => {
+      if (
+        preResponse.statusCode === 500 &&
+        request.route.options.tags.includes(ROUTE_TAG_AUTH_FLOW)
```

Comment:
I'm quite unsure about this check. It covers the use cases of SAML/OIDC when they fail on auth but I feel it's missing something.

Reply:
I think we'll need to confer with @azasypkin on this.

Reply:
@SiddharthMantri I reviewed this with Oleg today and I have a much better understanding now of what we're trying to achieve.

The tag check is appropriate, but ultimately we want to handle all error codes (status code >= 400) with some other specific conditions. The code block below @ ln219 handles the case for code 401 and checks if the login page is unavailable. I think we want to restructure/refactor this whole code block (registerOnPreResponse) to display the unauthenticated page if...
- we have any status code >= 400 AND
- the login page is unavailable OR
- we're running in serverless (e.g. the login page available but is not the default)

And of course, we should still fall through to other conditions if this is not the case.

It probably makes sense for us to sync on this when you have a chance.

cc: @azasypkin for visibility